cassiopeia/tools/branch_length_estimator/IIDExponentialMLE.py

```diff
@@ -0,0 +1,174 @@
+    solvers.
+
+    This estimator requires that the ancestral characters be provided (these
+    can be imputed with CassiopeiaTree's reconstruct_ancestral_characters
+    method if they are not known, which is usually the case for real data).
+
+    Because branch lengths and CRISPR/Cas9 mutation rate are not identifiable,
+    this estimator assumes that the tree has a depth of 1. In other words,
+    the estimated tree will have depth 1. The estimated mutation rate under
+    this unit-depth assumption will be stored as an attribute called
+    `mutation_rate`.
+
+    Missing states are treated as missing at random by the model.
+
+    Args:
+        minimum_branch_length: Estimated branch lengths will be constrained to
+            have length at least this value. By default it is set to 0.01,
+            since the MLE tends to collapse mutationless edges to length 0.
+        solver: Convex optimization solver to use. Can be "SCS", "ECOS", or
+            "MOSEK". Note that "MOSEK" solver should be installed separately.
+        verbose: Verbosity level.
+
+    Attributes:
+        log_likelihood: The log-likelihood of the training data under the
+            estimated model.
+        mutation_rate: The estimated CRISPR/Cas9 mutation rate, assuming that
+            the tree has depth exactly 1.
+    """
+
+    def __init__(
+        self,
+        minimum_branch_length: float = 0.01,
+        verbose: bool = False,
+        solver: str = "SCS",
+    ):
+        allowed_solvers = ["ECOS", "SCS", "MOSEK"]
+        if solver not in allowed_solvers:
+            raise ValueError(
+                f"Solver {solver} not allowed. Allowed solvers: {allowed_solvers}"
+            )  # pragma: no cover
+        self._minimum_branch_length = minimum_branch_length
+        self._verbose = verbose
+        self._solver = solver
+
+    def estimate_branch_lengths(self, tree: CassiopeiaTree) -> None:
+        r"""
+        See base class. The only caveat is that this method raises an
+        IIDExponentialMLEError if the underlying convex optimization
+        solver fails.
+
+        Raises:
+            IIDExponentialMLEError
+        """
+        # Extract parameters
+        minimum_branch_length = self._minimum_branch_length
+        solver = self._solver
+        verbose = self._verbose
+
+        # # # # # Create variables of the optimization problem # # # # #
+        r_X_t_variables = dict(
+            [
+                (node_id, cp.Variable(name=f"r_X_t_{node_id}"))
+                for node_id in tree.nodes
+            ]
+        )
+
+        # # # # # Create constraints of the optimization problem # # # # #
+        a_leaf = tree.leaves[0]
+        root = tree.root
+        root_has_time_0_constraint = [r_X_t_variables[root] == 0]
+        minimum_branch_length_constraints = [
+            r_X_t_variables[child]
+            >= r_X_t_variables[parent]
+            + minimum_branch_length * r_X_t_variables[a_leaf]
```

Comment:
I'm not sure I understand why you need to enforce that `time(child) >= time(parent) + min_branch_length + time(a_leaf)`. Specifically, why do you need to have the term `time(a_leaf)` here in this constraint? Thank you in advance for elucidating this!

In general, perhaps it would be good to make clear these constraints of the MLE problem.

Reply:
This is because the variables of the convex optimization problem are the times _multiplied_ by the mutation rate. Thus the following are equivalent, under the assumption that the tree has depth equal to 1:

1) `tree.get_time(child) >= tree.get_time(parent) + minimum_branch_length`

2) `r_X_t_variables[child] >= r_X_t_variables[parent] + minimum_branch_length * r_X_t_variables[a_leaf]`

The proof is: just divide by the mutation rate in (2), and use the fact that under the unit-depth assumption we have`r_X_t_variables[a_leaf] = mutation_rate`. You obtain exactly (1).

You might wonder why we need to parameterize the problem in terms of the time _multiplied_ by the rate, instead of just parameterizing the problem in terms of the times and the mutation rate parameter. This is because:
1) The times of the nodes and the mutation rate are not identifiable: they are only identifiable up to their product, so we don't gain anything with this richer parameterization (it has exactly 1 more parameter).
2) The optimization problem in terms of the times and the mutation rate parameter is (perhaps surprisingly) NOT a convex problem, whereas in terms of the times _multiplied_ by the mutation rate it is.

Note that all this is an implementation detail. What is promised by the class' docstring (that the MLE is computed assuming the tree has depth 1 and the branches have length at least `minimum_branch_length`) is exactly what is being done.

(Okay, technically there is at least one exception: when the character matrix is completely unmutated, in which case the MLE exists but I throw an error. This has to do with the fact what when the character matrix is totally unmutated, the MLE of the mutation rate is 0, and so the proof I gave above fails since we are dividing by zero! So: there actually are border cases where I don't do what is promised, but they are so extreme that nobody will ever encounter them in practice.)

Reply:
This makes a lot more sense, especially after reading your response to my question about how you set `mutation_rate`. Thanks again for walking me through this!